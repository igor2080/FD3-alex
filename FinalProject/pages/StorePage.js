import React, { Fragment } from 'react';

import StoreComponent from '../components/StoreComponent';
import StoreItem from '../components/StoreItem';
import { useParams, NavLink, withRouter, useHistory } from "react-router-dom";


var itemsArray = require('../SampleItems.json');

export default function StorePage() {
    const params = useParams();
    const history = useHistory();

    itemsArray = itemsArray.map(item => Object.assign(new StoreItem(), item));

    var pageNumber = parseInt(params.page);
    if (isNaN(pageNumber))
        pageNumber = 1;
    var from = (50 * pageNumber) - 49;
    var to = (50 * pageNumber);
    var maxPages = Math.ceil(itemsArray.length / 50);

    console.log("from:" + from + " to:" + to + ", max pages: " + maxPages);

    itemsArray = itemsArray.filter(item => item.itemId >= from && item.itemId < to);

    var toPreviousPage=()=>{
        history.push("/2");
    }

    return (
        <Fragment>
            <button onClick={toPreviousPage} disabled={pageNumber <= 1}>← Previous page</button>
            <button disabled={pageNumber >= maxPages}>Next page →</button>
            <StoreComponent storeItems={itemsArray} />
        </Fragment>
    );
}
