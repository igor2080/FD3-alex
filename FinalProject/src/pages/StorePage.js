import React from 'react';
import StoreComponent from '../components/StoreComponent';
import StoreItem from '../components/StoreItem';
import { NavLink, Link } from 'react-router-dom';
import { storeEvents } from '../components/events';

class StorePage extends React.PureComponent {
    state = {
        currentPageNumber: !isNaN(this.props.match.params.page) ? parseInt(this.props.match.params.page) : 1,
        itemsArray: [],
    }
    fakeDelay = 1000;//simulate loading times

    componentDidMount = () => {
        if (localStorage.getItem("itemArray") === null) {
            fetch('http://localhost:8080/sampleItems.json')
                .then(response => response.json())
                .then(data => this.loadData(data));
        }
        else {
            this.loadData(JSON.parse(localStorage.getItem("itemArray")));
        }
    }

    loadData = (data) => {
        var id = 0;
        this.setState({ itemsArray: data.map(item => Object.assign(new StoreItem(), item, { itemId: id++ })) });
        setTimeout(function () {//fake loading delay
            storeEvents.emit("dataReady");
        }, this.fakeDelay);

    }

    toPreviousPage = () => {
        this.setState((curState, props) => {
            return {
                currentPageNumber: (curState.currentPageNumber - 1),
            }
        });
    }

    toNextPage = () => {
        this.setState((curState, props) => {
            return {
                currentPageNumber: (curState.currentPageNumber + 1),
            }
        });
    }

    render() {
        var from = (50 * this.state.currentPageNumber) - 49;
        var to = (50 * this.state.currentPageNumber);
        var maxPages = Math.ceil(this.state.itemsArray.length / 50);
        var renderItemsArray = this.state.itemsArray.filter(item => item.itemId >= from && item.itemId <= to);
        return (
            <div>
                <NavLink to={{ pathname: "/" + this.state.currentPageNumber }} onClick={this.forceUpdate}>
                    <button onClick={this.toPreviousPage} disabled={this.state.currentPageNumber < 2}>← Previous page</button>
                </NavLink>

                <NavLink to={{ pathname: "/" + this.state.currentPageNumber }} onClick={this.forceUpdate}>
                    <button onClick={this.toNextPage} disabled={this.state.currentPageNumber >= maxPages}>Next page →</button>
                </NavLink>
                <StoreComponent storeItems={renderItemsArray} />
            </div>
        );
    }
}
export default StorePage;