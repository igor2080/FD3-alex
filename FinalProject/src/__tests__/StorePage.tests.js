"use strict"

import React from 'react';
import renderer from "react-test-renderer";
import { BrowserRouter } from 'react-router-dom';

import StorePage from '../pages/StorePage';
import { enableFetchMocks } from 'jest-fetch-mock'

test("StorePage to have an initial page count of 1", () => {

  const localStorageMock = {
    getItem: jest.fn(() => null),
    setItem: jest.fn(),
    clear: jest.fn()
  };
  global.localStorage = localStorageMock;
  enableFetchMocks()
  const component = renderer.create(
    <BrowserRouter>
      <StorePage match={{ params: { page: 1 } }} />
    </BrowserRouter>
  );
  let componentTree = component.toJSON();
  expect(componentTree).toMatchSnapshot();
})
