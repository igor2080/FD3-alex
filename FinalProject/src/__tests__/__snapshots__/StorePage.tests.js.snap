// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StorePage to have an initial page count of 1 1`] = `
<div>
  <a
    aria-current={null}
    className={undefined}
    href="/1"
    onClick={[Function]}
    style={undefined}
  >
    <button
      className="prevButton"
      disabled={true}
      onClick={[Function]}
    >
      ← Previous page
    </button>
  </a>
  <a
    aria-current={null}
    className={undefined}
    href="/1"
    onClick={[Function]}
    style={undefined}
  >
    <button
      disabled={true}
      onClick={[Function]}
    >
      Next page →
    </button>
  </a>
  <div>
    <label
      htmlFor="filterTextBox"
    >
      Filter: 
    </label>
    <input
      id="filterTextBox"
      onChange={[Function]}
      type="text"
    />
    <button
      onClick={[Function]}
    >
      Save to local storage
    </button>
    <button
      onClick={[Function]}
    >
      Reset local storage
    </button>
    <div
      className="box"
    >
      <div
        className="loader"
      />
    </div>
    <div
      className="flexdiv"
    >
      <span />
    </div>
  </div>
</div>
`;
